src/field_theory/separable_degree.lean

```diff
@@ -0,0 +1,166 @@
+/-
+Copyright (c) 2021 Jakob Scholbach. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob Scholbach
+-/
+import field_theory.separable
+import algebra.algebra.basic
+import data.polynomial.degree
+import algebra.char_p.exp_char
+
+/-!
+
+# Separable degree
+
+This file contains basics about the separable degree of a polynomial.
+
+## Main results
+
+- `is_separable_contraction`: is the condition that `g(x^(q^m)) = f(x)` for some `m : ℕ`
+- `has_separable_contraction`: the condition of having a separable contraction
+- `has_separable_contraction.degree`: the separable degree, defined as the degree of some
+  separable contraction
+- `irreducible_has_separable_contraction`: any nonzero irreducible polynomial can be contracted
+  to a separable polynomial
+- `has_separable_contraction.dvd_degree'`: the degree of a separable contraction divides the degree,
+  in function of the exponential characteristic of the field
+- `has_separable_contraction.dvd_degree` and `has_separable_contraction.eq_degree` specialize the
+  statement of `separable_degree_dvd_degree`
+- `is_separable_contraction.degree_eq`: the separable degree is well-defined, implemented as the
+  statement that the degree of any separable contraction equals `has_separable_contraction.degree`
+
+## Tags
+
+separable degree, degree, polynomial
+-/
+
+namespace polynomial
+
+noncomputable theory
+open_locale classical
+
+section comm_semiring
+
+variables {F : Type} [comm_semiring F] (q : ℕ)
+
+/-- A separable contraction of a polynomial `f` is a separable polynomial `g` such that
+`g(x^(q^m)) = f(x)` for some `m : ℕ`.-/
+def is_separable_contraction (f : polynomial F) (g : polynomial F) : Prop :=
+g.separable ∧ ∃ m : ℕ, expand F (q^m) g = f
+
+/-- The condition of having a separable contration. -/
+def has_separable_contraction (f : polynomial F) : Prop :=
+∃ g : polynomial F, is_separable_contraction q f g
+
+variables {q} {f : polynomial F} (hf : has_separable_contraction q f)
+
+/-- A choice of a separable contraction. -/
+def has_separable_contraction.contraction : polynomial F := classical.some hf
+
+/-- The separable degree of a polynomial is the degree of a given separable contraction. -/
+def has_separable_contraction.degree : ℕ := hf.contraction.nat_degree
+
+/-- The separable degree divides the degree, in function of the exponential characteristic of F. -/
+lemma has_separable_contraction.dvd_degree' :
```

Comment:
I think this lemma might be worth also stating for the `is_separable_contraction` version:
```
lemma is_separable_contraction.dvd_degree' {g} (hf : is_separable_contraction q f g) :
  ∃ m : ℕ, g.degree * (q ^ m) = f.nat_degree
```
Which will be all the non-classical bits of the proof below.

Reply:
OK, done.

Reply:
Done.